commit/chainfee/processor.go

```diff
@@ -3,41 +3,134 @@ package chainfee
+	}
 }
 
-func (w *Processor) Query(ctx context.Context, prevOutcome Outcome) (Query, error) {
+func (p *processor) Query(ctx context.Context, prevOutcome Outcome) (Query, error) {
 	return Query{}, nil
 }
 
-func (w *Processor) Observation(
+func (p *processor) Observation(
 	ctx context.Context,
 	prevOutcome Outcome,
 	query Query,
 ) (Observation, error) {
-	return Observation{}, nil
+	feeComponents := p.ccipReader.GetAvailableChainsFeeComponents(ctx)
+	nativeTokenPrices := p.ccipReader.GetWrappedNativeTokenPriceUSD(ctx, maps.Keys(feeComponents))
+	chainFeePriceUpdates := p.ccipReader.GetChainFeePriceUpdate(ctx, maps.Keys(feeComponents))
+	latestTimestamps := make(map[cciptypes.ChainSelector]time.Time, len(chainFeePriceUpdates))
+	for chain, update := range chainFeePriceUpdates {
+		latestTimestamps[chain] = update.Timestamp
+	}
+
+	return Observation{
+		FChain:                p.ObserveFChain(),
+		FeeComponents:         feeComponents,
+		NativeTokenPrices:     nativeTokenPrices,
+		ChainFeeLatestUpdates: latestTimestamps,
+		Timestamp:             time.Now().UTC(),
+	}, nil
 }
 
-func (w *Processor) Outcome(
+func (p *processor) Outcome(
 	prevOutcome Outcome,
 	query Query,
 	aos []plugincommon.AttributedObservation[Observation],
 ) (Outcome, error) {
-	return Outcome{}, nil
+
+	consensusObs, err := p.getConsensusObservation(aos)
+	if err != nil {
+		return Outcome{}, err
+	}
+	// No need to update yet
+	if len(consensusObs.FeeComponents) == 0 {
+		return Outcome{}, nil
+	}
+
+	// Stop early if earliest updated timestamp is still fresh
+	earliestUpdateTime := consensus.EarliestTimestamp(maps.Values(consensusObs.ChainFeeLatestUpdates))
+	nextUpdateTime := earliestUpdateTime.Add(p.ChainFeePriceBatchWriteFrequency.Duration())
+	if nextUpdateTime.After(consensusObs.Timestamp) {
+		return Outcome{}, nil
+	}
+
+	gasPrices := make([]cciptypes.GasPriceChain, 0, len(consensusObs.FeeComponents))
+	for chain, feeComp := range consensusObs.FeeComponents {
+		// GasPrice is a Bitwise operation here like:
+		// (dataAvFeeNative * nativeTokenPriceUSD) << 112 | (executionFeeNative * nativeTokenPrice)
+		// e.g. (dataAvFeeWei * WeiUSD) << 112 | (executionFeeWei * WeiUSD)
+		// nolint:lll
+		// https://github.com/smartcontractkit/chainlink/blob/60e8b1181dd74b66903cf5b9a8427557b85357ec/contracts/src/v0.8/ccip/FeeQuoter.sol#L498
+		nativeTokenPriceUSD := consensusObs.NativeTokenPrices[chain].Int
+
+		dataAvailabilityPriceUSD := new(big.Int).Mul(nativeTokenPriceUSD, feeComp.DataAvailabilityFee)
+		execPriceUSD := new(big.Int).Mul(nativeTokenPriceUSD, feeComp.ExecutionFee)
+
+		// Price is the combination ofr both dataAvailabilityPriceUSD and execPriceUSD
+		price := dataAvailabilityPriceUSD.Lsh(dataAvailabilityPriceUSD, 112)
+		combinedPricesUSD := new(big.Int).Or(price, execPriceUSD)
+
+		gasPrice := cciptypes.GasPriceChain{
+			ChainSel: chain,
+			GasPrice: cciptypes.NewBigInt(combinedPricesUSD),
+		}
+		gasPrices = append(gasPrices, gasPrice)
```

Comment:
This seems super EVM specific... how can it be generalized?

Reply:
Extracted into new structs that can later be changed